### PR TITLE
refactor(#2827): multi instance nvim-tree.live-filter

### LINKED ISSUE #2827: Multi Instance: Refactor: nvim-tree.live-filter
* #2255 

`nvim-tree.live-filter` is a stateful singleton. Make it an instantiable class owned by `Explorer`.

All usages must go through the `TreeExplorer` singleton.

### PR DESCRIPTION
**Pull request by raaymax**
fixes #2827 

This is not ready yet. Just putting it here to indicate that I'm working on it.
Also would be nice to have things from other PRs to not duplicate the work.

### CODE REVIEW

**Comment by alex-courtis** [comment]
Nice. I can see you on the board.

Added the issue in the description so that the PR also shows.

**Comment by alex-courtis** [comment]
Are you able to drag the issue on the board to "in progress"?

New to github projects and trying to sort out permissions...

**Comment by raaymax** [comment]
Inside the issue I can do almost everything but unfortunately project board is not interactive for me.

**Comment by alex-courtis** [comment]
> Inside the issue I can do almost everything but unfortunately project board is not interactive for me.

Not looking good for auto transitions: https://github.com/orgs/community/discussions/72699

I think I've added you this time; you should have full access to the project.

**Comment by raaymax** [comment]
yup, now it works

File: lua/nvim-tree/api.lua [hunk 1]
```diff
@@ -265,8 +264,18 @@ Api.git.reload = wrap(actions.reloaders.reload_git)
...
+  if explorer then
+    return explorer.live_filter:start_filtering(...)
+  end
+end)
+Api.live_filter.clear = wrap(function(...)
+  local explorer = core.get_explorer()
+  if explorer then
+    return explorer.live_filter:clear_filter(...)
+  end
+end)
```
Comment (alex-courtis) [inline lua/nvim-tree/api.lua:278]: Sorry, I didn't communicate this at all...

There's a somewhat hacky function to do this using member and method names:

```suggestion
Api.live_filter.start = wrap_explorer_member("live_filter", "start_filtering")
Api.live_filter.clear = wrap_explorer_member("live_filter", "clear_filter")
```

File: lua/nvim-tree/explorer/live-filter.lua [hunk 2]
```diff
@@ -2,17 +2,27 @@ local view = require "nvim-tree.view"
 local utils = require "nvim-tree.utils"
 local Iterator = require "nvim-tree.iterators.node-iterator"
 
-local M = {
-  filter = nil,
-}
+local LiveFilter = {}
```
Comment (alex-courtis) [inline lua/nvim-tree/explorer/live-filter.lua:5]: Could you please add luadoc here?

We can then put `---@field live_filter LiveFilter` into `@class Explorer`

That will allow autocompletion and (more importantly) language server validations/checks.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Looking great, works as expected!

 You are getting practiced at this.

File: lua/nvim-tree/explorer/live-filter.lua (hunk 2 above)
Comment (raaymax) [reply lua/nvim-tree/explorer/live-filter.lua:10]: I was wondering for what they are used for, thanks for clarification!

**Comment by raaymax** [comment]
For the next two weeks, I won't be available, but don't worry, I'll be back.

File: lua/nvim-tree/utils.lua [hunk 3]
```diff
@@ -112,7 +112,10 @@ function M.find_node(nodes, fn)
     end)
     :iterate()
   i = require("nvim-tree.view").is_root_folder_visible() and i or i - 1
-  i = require("nvim-tree.live-filter").filter and i + 1 or i
```
Comment (alex-courtis) [inline lua/nvim-tree/utils.lua:115]: Thank you.

Comment (alex-courtis) [inline lua/nvim-tree/utils.lua:115]: Thank you.

**Comment by alex-courtis** [comment]
> For the next two weeks, I won't be available, but don't worry, I'll be back.

No pressure at all. This is a slow burn...

**Review by alex-courtis — APPROVED** [review]
Many thanks for your contribution!

**Pull request merged**